src/services/codefixes/helpers.ts

```diff
@@ -417,6 +423,9 @@ export function createSignatureDeclarationFromSignature(
                     type = importableReference.typeNode;
                     importSymbols(importAdder, importableReference.symbols);
                 }
+                else if (!isDeclTypeSymbolVisible(parameterDecl.type)) {
+                    type = factory.createKeywordTypeNode(SyntaxKind.AnyKeyword);
+                }

```

Comment:
This is going to be too limited - it special-cases the parameter at the top level, but not other instances:

* deeper in the parameter type as a type argument

  ```ts
  foo(x: SomeType<TheInaccessibleType>): any
  ```

* deeper in the parameter type as the type of a property

  ```ts
  foo(x: { x: TheInaccessibleType }): any
  ```

* in the return type

  ```ts
  foo(): TheInaccessibleType
  ```

Here's my thought on the current change: you can fix the specific instance of the repro this way, but it's really a band-aid. I don't know how feasible it is, but I think a better fix would be one where we can somehow detect truncation when generating strings from type nodes.

I think others on the team will be able to give better pointers here.

Reply:
The symbol tracker actually has a `reportTruncationError` that is called so long as `NoTruncation` isn't passed. :P

You could stop passing `NoTruncation`, and if `reportTruncationError` on the SymbolTracker you pass in gets called, replace the whole generated type node with `any`.

Reply:
@DanielRosenwasser you're right, this does feel like a hack to handle this particular scenario. I'll continue working on it.

About the fourslash test, I actually wrote one but then I wasn't sure of how to test the completion result, since the given completion doesn't have an attached code action (and I only found ways to verify completion code actions) and the snippet property in the initial `completionInfo` was `undefined` (?). I might have missed something though.

Reply:
@weswigham well that was exactly the clue I needed :)

Reply:
Oh but wait, I _do_ need to pass `NoTruncation`:
```
if (context.truncating && context.flags & NodeBuilderFlags.NoTruncation) {
                context.tracker.reportTruncationError();
            }
```

Reply:
Is that a bug? Because then `checkTruncationLength` doesn't set `context.truncating` is the flag is set.

Reply:
If we don't report an error when you *do* pass the flag, then that's a bug. You *do* want to pass the `NoTruncation` flag.

Reply:
@DanielRosenwasser Why do I want to pass the flag? I want to use `defaultMaximumTruncationLength` as the bound in `checkTruncationLength`, instead of the super huge noTruncationMaximumTruncationLength`:
```
function checkTruncationLength(context: NodeBuilderContext): boolean {
    if (context.truncating) return context.truncating;
    return context.truncating = context.approximateLength > ((context.flags & NodeBuilderFlags.NoTruncation) ? noTruncationMaximumTruncationLength : defaultMaximumTruncationLength);
}
```

If I pass the flag, then `context.approximateLength` never exceeds `defaultMaximumTruncationLength` and the error isn't reported.

Reply:
You want `NoTruncation` because otherwise your type nodes can potentially be gibberish (filled in with invalid `...`

Reply:
@DanielRosenwasser but isn't the plan to replace them by `any` in that case?

Reply:
Type printing by default is too aggressive about replacing perfectly good types (because it was designed with error messages as the primary use-case). `NoTruncation` fixes that, and is usually what the language service should do.

Reply:
If the type is too big for an error message, it’s also too big to insert as a parameter type in a completion snippet. We shouldn’t be inserting tens of kilobytes of type printout that’s going to hang the user’s editor while they struggle to delete it.

Reply:
I won't have the chance in the next 2 days to figure out what the issue with (1) is - it sounds like you're saying it's not that the string is getting truncated, but rather, that the generated type nodes are getting reused in multiple parts of the tree, which the formatter **really** doesn't like.

My gut feel is that you want a combo of (2) and (3) - (2) is the golden path, and in some cases you want to really apply a truncation limit when the signature would be really bad. But I don't know how to thread that through.

Reply:
> I won't have the chance in the next 2 days to figure out what the issue with (1) is - it sounds like you're saying it's not that the string is getting truncated, but rather, that the generated type nodes are getting reused in multiple parts of the tree, which the formatter really doesn't like.

We looked at the situation of the crash under the debugger a bit as a group yesterday - what we saw in the debugger is that there's an off-by-one error somewhere in the formatters' synthetic node position assignment logic. A `}` that was supposed to be the child of a node was positioned one character paste the end of the node, so either the `}` had a slightly off position, or the node length for the parent was measured as one too short. I'm not too familiar with the position assignment logic the formatter uses for synthetic nodes, so I couldn't say which is more likely.